File: server-python/src/revturbine/core/state/impression_history.py
"""ImpressionHistory — Python port of @revt-eng/core/state/impression-history.ts.

Records and queries placement impression / interaction history; wraps any
``ImpressionHistoryStore``. Provides hot-path time-based
suppression caches for synchronous resolver checks.

Source: revturbine-scaffold/src/core/state/impression-history.ts
"""

from __future__ import annotations

import time
from datetime import datetime, timezone
from typing import Any

from revturbine.core.helpers import category_bucket
from revturbine.core.state.impression_history_types import (
    DEFAULT_DISMISS_COOLDOWN_MS,
    DEFAULT_SUPPRESSION_MS,
    ImpressionHistoryStore,
    ImpressionOutcome,
    ImpressionQuery,
    ImpressionRecord,
)

__all__ = [
    "DEFAULT_DISMISS_COOLDOWN_MS",
    "DEFAULT_SUPPRESSION_MS",
    "ImpressionHistory",
    "ImpressionHistoryOptions",
]


def _now_iso() -> str:
    """Match JS's ``new Date().toISOString()`` output (ms precision, ``Z`` suffix)."""
    return datetime.now(tz=timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def _suppress_until_iso(duration_ms: int) -> str:
    """Match JS's ``new Date(Date.now() + ms).toISOString()``."""
    target = datetime.fromtimestamp(time.time() + duration_ms / 1000, tz=timezone.utc)
    return target.isoformat(timespec="milliseconds").replace("+00:00", "Z")


class ImpressionHistoryOptions(dict[str, Any]):
    """Constructor argument bundle. Kept as a dict for parity with TS object
    literals; field access is via the keyword params on ``ImpressionHistory``."""


class ImpressionHistory:
    """Records and queries placement impression / interaction history.

    Source: impression-history.ts:29-239
    """

    def __init__(
        self,
        *,
        store: ImpressionHistoryStore,
        user_id: str,
        default_suppression_ms: int = DEFAULT_SUPPRESSION_MS,
        default_dismiss_cooldown_ms: int = DEFAULT_DISMISS_COOLDOWN_MS,
    ) -> None:
        self._store = store
        self._user_id = user_id
        self._default_suppression_ms = default_suppression_ms
        self._default_dismiss_cooldown_ms = default_dismiss_cooldown_ms
        self._explicit_suppressed_cache: dict[str, str] = {}
        self._suppressed_cache: dict[str, str] | None = None

    # ── Recording ──────────────────────────────────────────────────────────

    def record_impression(
        self,
        placement_id: str,
        payload_id: str | None = None,
        surface_template_id: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> None:
        """Source: impression-history.ts:55-62"""
        self._append_record(placement_id, "impressed", payload_id, surface_template_id, metadata)

    def record_dismissal(
        self,
        placement_id: str,
        payload_id: str | None = None,
        surface_template_id: str | None = None,
        metadata: dict[str, Any] | None = None,
        cooldown_ms: int | None = None,
    ) -> None:
        """Records a dismissal. Time-boxed: the placement is hidden for the
        dismiss cooldown (``cooldown_after_dismiss_days``, default 7 days) and
        re-shows once it elapses. NOT permanent (plan 167, Q-1).

        Source: impression-history.ts:68-76
        """
        ms = cooldown_ms if cooldown_ms is not None else self._default_dismiss_cooldown_ms
        suppress_until = _suppress_until_iso(ms)
        merged_metadata: dict[str, Any] = {**(metadata or {}), "suppressUntil": suppress_until}
        self._append_record(
            placement_id, "dismissed", payload_id, surface_template_id, merged_metadata
        )
        self._suppress_in_cache(placement_id, suppress_until)

    def record_click_thru(
        self,
        placement_id: str,
        payload_id: str | None = None,
        surface_template_id: str | None = None,
        metadata: dict[str, Any] | None = None,
        cooldown_ms: int | None = None,
    ) -> None:
        """Records a bare click-through — clicked but not confirmed complete
        (e.g. abandoned checkout). Treated as a dismiss cooldown; the placement
        may return. For a confirmed conversion use ``record_conversion``
        (plan 167, Q-1).

        Source: impression-history.ts:82-90
        """
        ms = cooldown_ms if cooldown_ms is not None else self._default_dismiss_cooldown_ms
        suppress_until = _suppress_until_iso(ms)
        merged_metadata: dict[str, Any] = {**(metadata or {}), "suppressUntil": suppress_until}
        self._append_record(
            placement_id, "clicked_thru", payload_id, surface_template_id, merged_metadata
        )
        self._suppress_in_cache(placement_id, suppress_until)

    def record_conversion(
        self,
        placement_id: str,
        payload_id: str | None = None,
        surface_template_id: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> None:
        """Records conversion analytics without changing eligibility (plan 254).

        Source: impression-history.ts:record_conversion
        """
        self._append_record(
            placement_id, "cta_completed", payload_id, surface_template_id, metadata
        )

    def record_suppression(
        self,
        placement_id: str,
        payload_id: str | None = None,
        surface_template_id: str | None = None,
        metadata: dict[str, Any] | None = None,
        duration_ms: int | None = None,
    ) -> None:
        """Records a time-based suppression. The ``suppressUntil`` timestamp
        is appended to ``metadata`` so downstream queries can recover the
        window without external context.

        Source: impression-history.ts:98-112
        """
        ms = duration_ms if duration_ms is not None else self._default_suppression_ms
        suppress_until = _suppress_until_iso(ms)
        merged_metadata: dict[str, Any] = {**(metadata or {}), "suppressUntil": suppress_until}
        self._append_record(
            placement_id, "suppressed", payload_id, surface_template_id, merged_metadata
        )
        self._suppress_in_cache(placement_id, suppress_until)
        self._explicit_suppressed_cache[placement_id] = suppress_until

    # ── Querying ───────────────────────────────────────────────────────────

    def is_retired(self, placement_id: str) -> bool:
        """Compatibility method: conversion no longer retires placements."""
        return False

    def get_retired_ids(self) -> set[str]:
        """Compatibility method: no interaction permanently retires a placement."""
        return set()

    def is_retired_sync(self, placement_id: str) -> bool:
        """Compatibility method: conversion no longer retires placements."""
        return False

    def is_suppressed_sync(self, placement_id: str) -> bool:
        """Synchronous check for time-based suppression. Expired entries are
        evicted from the cache as a side effect.

        Source: impression-history.ts:149-157
        """
        if self._suppressed_cache is None:
            return False
        until = self._suppressed_cache.get(placement_id)
        if until is None:
            return False
        if _parse_iso_to_ms(until) > int(time.time() * 1000):
            return True
        del self._suppressed_cache[placement_id]
        return False

    def is_hidden_sync(self, placement_id: str, category: str | None = None) -> bool:
        """Fixed/Gated bypass user cooldowns, retaining explicit suppression."""
        until = self._explicit_suppressed_cache.get(placement_id)
        if until and _parse_iso_to_ms(until) > int(time.time() * 1000):
            return True
        self._explicit_suppressed_cache.pop(placement_id, None)
        if category_bucket(category or "") <= 1:
            return False
        return self.is_suppressed_sync(placement_id)

    def query_history(self, query: ImpressionQuery | None = None) -> list[ImpressionRecord]:
        """Source: impression-history.ts:169-171"""
        return self._store.query(self._user_id, query)

    # ── Lifecycle ──────────────────────────────────────────────────────────

    def hydrate(self) -> None:
        """Pre-warm the timed suppression caches from the store. Call
        during SDK initialization for synchronous access.

        Source: impression-history.ts:181-184
        """
        self._suppressed_cache = self._store.get_suppressed_placements(self._user_id)
        self._explicit_suppressed_cache = {}
        for record in self._store.query(self._user_id, ImpressionQuery(outcomes=["suppressed"])):
            until = (record.get("metadata") or {}).get("suppressUntil")
            if (
                record["outcome"] == "suppressed"
                and isinstance(until, str)
                and _parse_iso_to_ms(until) > int(time.time() * 1000)
            ):
                self._explicit_suppressed_cache.setdefault(record["placement_id"], until)

    def reset(self) -> None:
        """Clear all impression history for this user.

        Source: impression-history.ts:189-193
        """
        self._store.clear(self._user_id)
        self._explicit_suppressed_cache.clear()
        self._suppressed_cache = {}

    def set_user_id(self, user_id: str) -> None:
        """Switch user identity — clears caches and points at a new user.

        Source: impression-history.ts:198-202
        """
        self._user_id = user_id
        self._explicit_suppressed_cache.clear()
        self._suppressed_cache = None

    # ── Internal ───────────────────────────────────────────────────────────

    def _append_record(
        self,
        placement_id: str,
        outcome: ImpressionOutcome,
        payload_id: str | None,
        surface_template_id: str | None,
        metadata: dict[str, Any] | None,
    ) -> None:
        record: ImpressionRecord = ImpressionRecord(
            placement_id=placement_id,
            outcome=outcome,
            occurred_at=_now_iso(),
        )
        if payload_id:
            record["payload_id"] = payload_id
        if surface_template_id:
            record["surface_template_id"] = surface_template_id
        if metadata:
            record["metadata"] = metadata
        self._store.append(self._user_id, record)

    def _suppress_in_cache(self, placement_id: str, suppress_until: str) -> None:
        if self._suppressed_cache is None:
            self._suppressed_cache = {}
        self._suppressed_cache[placement_id] = suppress_until


def _parse_iso_to_ms(iso: str) -> int:
    """Mirror JS ``new Date(iso).getTime()``. Tolerates trailing ``Z``."""
    try:
        normalized = iso.replace("Z", "+00:00") if iso.endswith("Z") else iso
        return int(datetime.fromisoformat(normalized).timestamp() * 1000)
    except (TypeError, ValueError):
        return 0

File: server-rust/src/state/impression_history.rs
//! Records and queries placement impression history.
//!
//! Wraps any [`ImpressionHistoryStore`] and maintains hot-path caches so the
//! resolver can ask "is this hidden?" synchronously.
//!
//! Source: revturbine-scaffold/src/core/state/impression-history.ts

use std::collections::{HashMap, HashSet};

use crate::placements::selection::category_bucket;
use serde_json::{json, Value};

use crate::state::impression_history_types::{
    iso_from_ms, parse_iso_to_ms, ImpressionHistoryStore, ImpressionQuery, ImpressionRecord,
    DEFAULT_DISMISS_COOLDOWN_MS, DEFAULT_SUPPRESSION_MS,
};

fn system_now_ms() -> i64 {
    use std::time::{SystemTime, UNIX_EPOCH};
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0)
}

/// Records and queries placement impression / interaction history.
///
/// # Cache semantics
///
/// `retired` and `suppressed` caches start **cold** (`None`), not empty. The
/// `*_sync` checks return `false` on a cold cache rather than hitting the
/// store — they are hot-path calls on the resolver, and a cold cache means
/// "not yet known", not "nothing hidden". Call [`Self::hydrate`] during
/// initialization to warm them.
///
/// Source: impression-history.ts:29-239
pub struct ImpressionHistory<S: ImpressionHistoryStore> {
    store: S,
    user_id: String,
    default_suppression_ms: i64,
    default_dismiss_cooldown_ms: i64,
    retired_cache: Option<HashSet<String>>,
    suppressed_cache: Option<HashMap<String, String>>,
    explicit_suppressed_cache: HashMap<String, String>,
    now_fn: Box<dyn Fn() -> i64 + Send + Sync>,
}

impl<S: ImpressionHistoryStore> ImpressionHistory<S> {
    /// Construct with default windows and the system clock.
    pub fn new(store: S, user_id: &str) -> Self {
        Self::with_options(
            store,
            user_id,
            DEFAULT_SUPPRESSION_MS,
            DEFAULT_DISMISS_COOLDOWN_MS,
            Box::new(system_now_ms),
        )
    }

    /// Construct with explicit windows and clock.
    pub fn with_options(
        store: S,
        user_id: &str,
        default_suppression_ms: i64,
        default_dismiss_cooldown_ms: i64,
        now_fn: Box<dyn Fn() -> i64 + Send + Sync>,
    ) -> Self {
        Self {
            store,
            user_id: user_id.to_string(),
            default_suppression_ms,
            default_dismiss_cooldown_ms,
            retired_cache: None,
            suppressed_cache: None,
            explicit_suppressed_cache: HashMap::new(),
            now_fn,
        }
    }

    // ── Recording ───────────────────────────────────────────────────────────

    /// Record that the placement was shown. Carries no suppression.
    ///
    /// Source: impression-history.ts:55-62
    pub fn record_impression(&mut self, placement_id: &str, metadata: Option<Value>) {
        self.append_record(placement_id, "impressed", metadata);
    }

    /// Record a dismissal — **time-boxed**, not permanent. The placement
    /// returns once the cooldown elapses (plan 167 Q-1).
    ///
    /// Source: impression-history.ts:68-76
    pub fn record_dismissal(
        &mut self,
        placement_id: &str,
        metadata: Option<Value>,
        cooldown_ms: Option<i64>,
    ) {
        let ms = cooldown_ms.unwrap_or(self.default_dismiss_cooldown_ms);
        self.append_suppressing_record(placement_id, "dismissed", metadata, ms);
    }

    /// Record a bare click-through — clicked but not confirmed complete (an
    /// abandoned checkout, say). Treated as a dismiss cooldown; the placement
    /// may return. Use [`Self::record_conversion`] for a confirmed conversion.
    ///
    /// Source: impression-history.ts:82-90
    pub fn record_click_thru(
        &mut self,
        placement_id: &str,
        metadata: Option<Value>,
        cooldown_ms: Option<i64>,
    ) {
        let ms = cooldown_ms.unwrap_or(self.default_dismiss_cooldown_ms);
        self.append_suppressing_record(placement_id, "clicked_thru", metadata, ms);
    }

    /// Record conversion analytics without changing placement eligibility.
    ///
    /// Source: impression-history.ts (recordConversion)
    pub fn record_conversion(&mut self, placement_id: &str, metadata: Option<Value>) {
        self.append_record(placement_id, "cta_completed", metadata);
    }

    /// Record a time-based suppression.
    ///
    /// Source: impression-history.ts:98-112
    pub fn record_suppression(
        &mut self,
        placement_id: &str,
        metadata: Option<Value>,
        duration_ms: Option<i64>,
    ) {
        let ms = duration_ms.unwrap_or(self.default_suppression_ms);
        self.append_suppressing_record(placement_id, "suppressed", metadata, ms);
    }

    // ── Querying ────────────────────────────────────────────────────────────

    /// Compatibility method: conversion no longer retires placements.
    ///
    /// Source: impression-history.ts:121-124
    pub fn is_retired(&mut self, _placement_id: &str) -> bool {
        false
    }

    /// Compatibility query: conversions no longer retire placements.
    pub fn get_retired_ids(&mut self) -> &HashSet<String> {
        self.retired_cache.get_or_insert_with(HashSet::new)
    }

    /// Compatibility query: conversions no longer retire placements.
    #[must_use]
    pub fn is_retired_sync(&self, _placement_id: &str) -> bool {
        false
    }

    /// Synchronous suppression check. Expired entries are **evicted** as a
    /// side effect, so a lapsed cooldown stops costing a lookup.
    ///
    /// Source: impression-history.ts:149-157
    pub fn is_suppressed_sync(&mut self, placement_id: &str) -> bool {
        let now = (self.now_fn)();
        let Some(cache) = self.suppressed_cache.as_mut() else {
            return false;
        };
        let Some(until) = cache.get(placement_id) else {
            return false;
        };
        if parse_iso_to_ms(until) > now {
            return true;
        }
        cache.remove(placement_id);
        false
    }

    /// Whether the placement has an active timed suppression.
    ///
    /// Source: impression-history.ts:162-164
    pub fn is_hidden_sync(&mut self, placement_id: &str) -> bool {
        self.is_hidden_for_category_sync(placement_id, "")
    }

    /// Category-aware visibility: explicit suppression applies to every category.
    pub fn is_hidden_for_category_sync(&mut self, placement_id: &str, category: &str) -> bool {
        let explicit = self
            .explicit_suppressed_cache
            .get(placement_id)
            .is_some_and(|until| parse_iso_to_ms(until) > (self.now_fn)());
        if explicit {
            return true;
        }
        self.explicit_suppressed_cache.remove(placement_id);
        if category_bucket(category) <= 1 {
            return false;
        }
        self.is_suppressed_sync(placement_id)
    }

    /// Query the underlying store.
    ///
    /// Source: impression-history.ts:169-171
    pub fn query_history(&self, query: Option<&ImpressionQuery>) -> Vec<ImpressionRecord> {
        self.store.query(&self.user_id, query)
    }

    // ── Lifecycle ───────────────────────────────────────────────────────────

    /// Pre-warm both caches. Call during initialization so the `*_sync` checks
    /// are meaningful.
    ///
    /// Source: impression-history.ts:181-184
    pub fn hydrate(&mut self) {
        self.retired_cache = Some(HashSet::new());
        self.suppressed_cache = Some(self.store.get_suppressed_placements(&self.user_id));
        self.explicit_suppressed_cache.clear();
        let query = ImpressionQuery {
            outcomes: Some(vec!["suppressed".into()]),
            ..Default::default()
        };
        for record in self.store.query(&self.user_id, Some(&query)) {
            if record.outcome != "suppressed" {
                continue;
            }
            if let Some(until) = record
                .metadata
                .as_ref()
                .and_then(|m| m.get("suppressUntil"))
                .and_then(Value::as_str)
            {
                if parse_iso_to_ms(until) > (self.now_fn)() {
                    self.explicit_suppressed_cache
                        .entry(record.placement_id)
                        .or_insert_with(|| until.to_string());
                }
            }
        }
    }

    /// Clear all history for this user. Caches become empty (warm), not cold.
    ///
    /// Source: impression-history.ts:189-193
    pub fn reset(&mut self) {
        self.store.clear(&self.user_id);
        self.retired_cache = Some(HashSet::new());
        self.suppressed_cache = Some(HashMap::new());
        self.explicit_suppressed_cache.clear();
    }

    /// Switch user identity. Caches go **cold**, not empty — the new user's
    /// history is unknown, and treating it as empty would leak a "nothing
    /// hidden" answer for someone whose history has not been read.
    ///
    /// Source: impression-history.ts:198-202
    pub fn set_user_id(&mut self, user_id: &str) {
        self.user_id = user_id.to_string();
        self.retired_cache = None;
        self.suppressed_cache = None;
        self.explicit_suppressed_cache.clear();
    }

    // ── Internal ────────────────────────────────────────────────────────────

    fn append_record(&mut self, placement_id: &str, outcome: &str, metadata: Option<Value>) {
        let mut record =
            ImpressionRecord::new(placement_id, outcome, &iso_from_ms((self.now_fn)()));
        record.metadata = metadata;
        self.store.append(&self.user_id, record);
    }

    /// Append a record whose metadata carries a `suppressUntil`, and mirror
    /// the window into the cache so the next `*_sync` call sees it without a
    /// store round-trip.
    fn append_suppressing_record(
        &mut self,
        placement_id: &str,
        outcome: &str,
        metadata: Option<Value>,
        duration_ms: i64,
    ) {
        let suppress_until = iso_from_ms((self.now_fn)() + duration_ms);
        let mut merged = match metadata {
            Some(Value::Object(m)) => Value::Object(m),
            _ => json!({}),
        };
        merged["suppressUntil"] = Value::String(suppress_until.clone());

        self.append_record(placement_id, outcome, Some(merged));
        if outcome == "suppressed" {
            self.explicit_suppressed_cache
                .insert(placement_id.to_string(), suppress_until.clone());
        }
        self.suppressed_cache
            .get_or_insert_with(HashMap::new)
            .insert(placement_id.to_string(), suppress_until);
    }
}
